File: rust/src/equipment/relics.rs
use crate::equipment::latex_table::{TableRow, ToTableRows};
use crate::equipment::{
    all_apparel, all_implements, all_magic_armor, all_magic_weapons, all_tools, latex_table,
    Apparel, Implement, ItemRarity, MagicArmor, MagicWeapon, Tool,
};

enum Relic {
    Apparel(Apparel),
    Implement(Implement),
    MagicArmor(MagicArmor),
    MagicWeapon(MagicWeapon),
    Tool(Tool),
}

impl Relic {
    fn to_latex(&self) -> String {
        match self {
            Self::Apparel(i) => i.to_latex(),
            Self::Implement(i) => i.to_latex(),
            Self::MagicArmor(i) => i.to_latex(),
            Self::MagicWeapon(i) => i.to_latex(),
            Self::Tool(i) => i.clone().to_latex(),
        }
    }
}

impl ToTableRows for Relic {
    fn to_table_rows(&self) -> Vec<TableRow> {
        match self {
            Self::Apparel(i) => i.to_table_rows(),
            Self::Implement(i) => i.to_table_rows(),
            Self::MagicArmor(i) => i.to_table_rows(),
            Self::MagicWeapon(i) => i.to_table_rows(),
            Self::Tool(i) => i.to_table_rows(),
        }
    }
}

fn all_relics() -> Vec<Relic> {
    let mut relics = vec![];

    relics.append(
        &mut all_apparel(Some(ItemRarity::Relic))
            .into_iter()
            .map(Relic::Apparel)
            .collect(),
    );
    relics.append(
        &mut all_implements(Some(ItemRarity::Relic))
            .into_iter()
            .map(Relic::Implement)
            .collect(),
    );
    relics.append(
        &mut all_magic_armor(Some(ItemRarity::Relic))
            .into_iter()
            .map(Relic::MagicArmor)
            .collect(),
    );
    relics.append(
        &mut all_magic_weapons(Some(ItemRarity::Relic))
            .into_iter()
            .map(Relic::MagicWeapon)
            .collect(),
    );
    relics.append(
        &mut all_tools(None, Some(ItemRarity::Relic))
            .into_iter()
            .map(Relic::Tool)
            .collect(),
    );

    relics
}

pub fn all_relic_descriptions() -> Vec<String> {
    all_relics().iter().map(|r| r.to_latex()).collect()
}

pub fn relics_table() -> String {
    let with_category = true;

    let mut rows: Vec<TableRow> = all_relics()
        .iter()
        .map(|r| r.to_table_rows())
        .flatten()
        .collect();
    latex_table::standard_sort(&mut rows);

    latex_table::longtable("Relics", rows, with_category)
}

File: typescript/src/equipment/latex/relics.ts
import { allMagicArmor } from '../data/magic_armor';
import { allMagicWeapons } from '../data/magic_weapons';
import { allImplements } from '../data/implements';
import { allApparel, apparelLatex } from '../apparel';
import { allTools } from '../data/all_tools';
import { itemLatex } from './item_latex';
import * as table from './latex_table';
import { ToolCategory } from '../types';
import { getToolCategoryName, getToolCraftingLatex } from './tools';

export function generateRelicDescriptions(): string {
  const armor = allMagicArmor()
    .filter((m) => m.item.rarity === 'Relic')
    .map((m) => {
      const materials = m.kind === 'Body' ? 'bone, leather, or metal' : 'bone, metal, or wood';
      const category = m.kind === 'Body' ? 'Body armor' : 'Shield';
      return itemLatex(m.item, `${category} -- Craft (${materials})`);
    });

  const weapons = allMagicWeapons()
    .filter((m) => m.item.rarity === 'Relic')
    .map((m) => itemLatex(m.item, 'Craft (as base weapon)'));

  const apparel = allApparel('Relic').map((a) => apparelLatex(a));

  const implementsData = allImplements()
    .filter((m) => m.item.rarity === 'Relic')
    .map((m) => {
      let craftingLatex = '';
      if (m.kind === 'Staff') craftingLatex = 'bone or wood';
      else if (m.kind === 'Rod') craftingLatex = 'bone, metal, or wood';
      else if (m.kind === 'Wand') craftingLatex = 'bone or wood';
      return itemLatex(m.item, `${m.kind} -- ${craftingLatex}`);
    });

  const tools = allTools({ rarity: 'Relic' }).map((t) =>
    itemLatex(t.item, getToolCraftingLatex(t.category)),
  );

  const all = [...apparel, ...implementsData, ...armor, ...weapons, ...tools];
  return all.join('\n');
}

export function generateRelicsTable(): string {
  const rows: table.TableRow[] = [];

  rows.push(
    ...allApparel('Relic').flatMap((a) => table.fromItem(a.item, false, a.kind)),
  );
  rows.push(
    ...allMagicArmor()
      .filter((m) => m.item.rarity === 'Relic')
      .flatMap((m) => table.fromItem(m.item, false, m.kind === 'Body' ? 'Body armor' : 'Shield')),
  );
  rows.push(
    ...allMagicWeapons()
      .filter((m) => m.item.rarity === 'Relic')
      .flatMap((m) => table.fromItem(m.item, false, 'Weapon')),
  );
  rows.push(
    ...allImplements()
      .filter((m) => m.item.rarity === 'Relic')
      .flatMap((m) => table.fromItem(m.item, false, m.kind)),
  );
  rows.push(
    ...allTools({ rarity: 'Relic' }).flatMap((t) =>
      table.fromItem(t.item, isConsumable(t), getToolCategoryName(t.category)),
    ),
  );

  table.standardSort(rows);
  return table.longtable('Relics', rows, true);
}

function isConsumable(tool: any): boolean {
  if (typeof tool.category === 'string') {
    return (
      tool.category === 'Alchemical' || tool.category === 'Poison' || tool.category === 'Potion'
    );
  }
  return false;
}
